force-app/main/releaseGating/classes/ReleaseGateVMapper.cls

```diff
@@ -0,0 +1,126 @@
+     * @description Instance for Singleton Pattern
+     **********************************************************************************/
+    @TestVisible
+    private static ReleaseGateVMapper instance;
+
+    /**********************************************************************************
+     * @description Empty constructor is not public for Singleton Pattern
+     **********************************************************************************/
+    protected ReleaseGateVMapper() {
+    }
+
+    /**********************************************************************************
+     * @description Static method to get the current instance for Singleton pattern
+     * @return The instance of ReleaseGateVMapper.
+     **********************************************************************************/
+    public static ReleaseGateVMapper getInstance() {
+        if (instance == null) {
+            instance = new ReleaseGateVMapper();
+        }
+
+        return instance;
+    }
+
+    /**********************************************************************************************************
+     * @description Retrieves a list of ReleaseGateVModel for the specified product registry
+     * @param  productRegistryName product registry developer name
+     * @return A list of ReleaseGateVModel
+     *********************************************************************************************************/
+    public virtual List<ReleaseGateVModel> getReleaseGateVModel(String productRegistryName) {
+        GetReleaseGateActionResultModel releaseGateModel = locateReleaseGateService()
+            .getReleaseGateModel(productRegistryName);
+        return this.mapReleaseGateVModel(releaseGateModel);
+    }
+
+    /**
+     * @description maps GetReleaseGateActionResultModel to List<ReleaseGateVModel>
+     * @param  getReleaseGateModel source GetReleaseGateActionResultModel instance
+     * @return                     a list of ReleaseGateVModel instances
+     */
+    protected virtual List<ReleaseGateVModel> mapReleaseGateVModel(
+        GetReleaseGateActionResultModel getReleaseGateModel
+    ) {
+        List<ReleaseGateVModel> releaseGateVModelList = new List<ReleaseGateVModel>();
+        for (GetReleaseGateActionResultModel.GetReleaseGateResult releaseGateResult : getReleaseGateModel.result) {
+            ReleaseGateVModel.ReleaseGateProduct releaseProduct = new ReleaseGateVModel.ReleaseGateProduct(
+                releaseGateResult.product.name,
+                releaseGateResult.product.label,
+                releaseGateResult.product.nextReleaseDate
+            );
+            List<ReleaseGateVModel.ReleaseGate> gates = new List<ReleaseGateVModel.ReleaseGate>();
+            for (GetReleaseGateActionResultModel.ReleaseGate releaseGate : releaseGateResult.gates) {
+                List<ReleaseGateVModel.ReleaseGateFeature> features = new List<ReleaseGateVModel.ReleaseGateFeature>();
```

Comment:
Consider extracting this into a new method for testability. Any time we have a loop in a loop, it makes me feel like we should look into extracting the method. In this case, we may want to extract two.

Reply:
Refactored into 3 methods